### ISSUE TITLE
Code and data for LID model

### ISSUE THREAD

**Issue by skomban** [facebookresearch/fairseq#4561]
Hi, my question is around Language Identification (LID) part of NLLB and was wondering if the code and data used to train the LID model are part of this codebase? My goal is to first reproduce the results from the paper and then later train on my own data. Any help would be much appreciated. Thanks.

**Comment by Celebio** [comment]
hi @skomban,
we don't plan to publish the data we used for training the LID model. Regarding the training code, we used regular [fastText supervised](https://fasttext.cc/docs/en/supervised-tutorial.html) training. There is no plan to open-source the preprocessing steps, but you can refer to the section 5.1.3 of our [paper](https://research.facebook.com/publications/no-language-left-behind/): "Filtering Training Data". Script detection used here corresponds to [this](https://github.com/facebookresearch/stopes/blob/main/stopes/pipelines/monolingual/utils/predict_script.py) open-sourced code.

Regards,
Onur

**Comment by skomban** [comment]
Hi Onur, thank you for your prompt response. One further question, any plans on open sourcing the LID Inspection Interface? If not, could you give me some pointers on how I could do it myself. Right now if I understand correctly the fastText api does not support querying the ngram weights, right? Thanks.

**Comment by skomban** [comment]
Hi @Celebio,
Thank you for your prompt response. One further question, any plans on open sourcing the LID Inspection Interface? If not, could you give me some pointers on how I could do it myself. Right now if I understand correctly the fastText api does not support querying the ngram weights, right? 
Thanks,
Stanley

**Comment by Celebio** [comment]
hi @skomban,
On the high level, it computes the dot product of each subword's vector with each label's vector. Then we normalize for each character (since a character appears in several subwords), which gives the character "heatmap".

You are right, the public fastText api doesn't have the `explain` method. It is possible that we publish it at some point but we can't commit to a specific date. Thank you for letting us know that it could be useful to the community.

Regards,
Onur

**Comment by skomban** [comment]
Hi @Celebio,
Thank you again for your prompt reply and for all the clarifications. The _explain_ method would certainly be useful particularly for debugging. Fingers crossed that it soon becomes part of the fastText api.

Thanks,
Stanley
